packages/x-date-pickers-pro/src/DateRangeCalendar/DateRangeCalendar.tsx

```diff
@@ -0,0 +1,553 @@
+    // In both cases, we don't want to change the visible month(s).
+    if (disableAutoMonthSwitching && prevDate && utils.isEqual(prevDate, date)) {
+      return;
+    }
+
+    const displayingMonthRange = calendars - 1;
+    const currentMonthNumber = utils.getMonth(calendarState.currentMonth);
+    const requestedMonthNumber = utils.getMonth(date);
+
+    if (
+      !utils.isSameYear(calendarState.currentMonth, date) ||
+      requestedMonthNumber < currentMonthNumber ||
+      requestedMonthNumber > currentMonthNumber + displayingMonthRange
+    ) {
+      const newMonth =
+        currentDatePosition === 'start'
+          ? date
+          : // If need to focus end, scroll to the state when "end" is displaying in the last calendar
+            utils.addMonths(date, -displayingMonthRange);
+
+      changeMonth(newMonth);
+    }
+  }, [currentDatePosition, value]); // eslint-disable-line
+
+  const handleSelectedDayChange = useEventCallback((newDate: TDate | null) => {
+    const { nextSelection, newRange } = calculateRangeChange({
+      newDate,
+      utils,
+      range: value,
+      currentlySelectingRangeEnd: currentDatePosition,
+    });
+
+    setCurrentDatePosition(nextSelection);
+    onCurrentDatePositionChange?.(nextSelection);
+
+    const isFullRangeSelected = currentDatePosition === 'end' && isRangeValid(utils, newRange);
+
+    setValue(newRange);
+    onChange?.(newRange, isFullRangeSelected ? 'finish' : 'partial');
+  });
+
+  const selectNextMonth = React.useCallback(() => {
+    changeMonth(utils.getNextMonth(calendarState.currentMonth));
+  }, [changeMonth, calendarState.currentMonth, utils]);
+
+  const selectPreviousMonth = React.useCallback(() => {
+    changeMonth(utils.getPreviousMonth(calendarState.currentMonth));
+  }, [changeMonth, calendarState.currentMonth, utils]);
+
+  const isNextMonthDisabled = useNextMonthDisabled(calendarState.currentMonth, {
+    disableFuture,
+    maxDate,
+  });
+
+  const isPreviousMonthDisabled = usePreviousMonthDisabled(calendarState.currentMonth, {
+    disablePast,
+    minDate,
+  });
+
+  const baseDateValidationProps: Required<BaseDateValidationProps<TDate>> = {
+    disablePast,
+    disableFuture,
+    maxDate,
+    minDate,
+  };
+
+  const commonViewProps = {
+    disableHighlightToday,
+    readOnly,
+    disabled,
+  };
+
+  // When disabled, limit the view to the selected date
+  const minDateWithDisabled = (disabled && value[0]) || minDate;
+  const maxDateWithDisabled = (disabled && value[1]) || maxDate;
+
+  const [rangePreviewDay, setRangePreviewDay] = React.useState<TDate | null>(null);
+
+  const CalendarTransitionProps = React.useMemo(
+    () => ({
+      onMouseLeave: () => setRangePreviewDay(null),
+    }),
+    [],
+  );
+
+  // Range going for the start of the start day to the end of the end day.
+  // This makes sure that `isWithinRange` works with any time in the start and end day.
```

Comment:
Would it make sense to remove that part and replace `isWithinRange` with a helper `isWithinDayRange` which does the check by ignoring the time

Reply:
It is what I did at first in https://github.com/mui/mui-x/pull/6403
But we discussed with @LukasTy and that first draft was doing the date manipulation for everyday instead of once in the render.
So that's for performance reasons.